wave_lang/kernel/wave/barriers.py

```diff
@@ -113,33 +134,93 @@ def add_shared_memory_barriers(
         if mem := is_shared_memory_op(custom):
             state = info[mem]
             if state.last_node and need_barrier(custom, state.last_node):
-                if barrier := is_barrier_between(
-                    state.last_node.fx_node, custom.fx_node
-                ):
+                if barrier := is_barrier_between(state.last_node.fx_node, node):
                     barrier = get_custom(barrier)
-                    # Promote the barrier to wait for async ops
-                    if state.is_async and not barrier.wait_async_ops:
-                        barrier.update_arg("wait_async_ops", True)
+                    # Add async deps to the barrier.
+                    if state.async_deps:
+                        deps = list(
+                            dict.fromkeys(barrier.async_deps)
+                            | dict.fromkeys(state.async_deps)
+                        )
+                        barrier.update_arg("async_deps", deps)
                 else:
                     # Synchronize after the write to shared memory before we read from it.
+                    deps = list(dict.fromkeys(state.async_deps))
                     with graph.inserting_before(node):
-                        SharedMemoryBarrier(wait_async_ops=state.is_async).add_to_graph(
-                            graph
-                        )
+                        SharedMemoryBarrier(async_deps=deps).add_to_graph(graph)
 
-                state.is_async = False
+                state.async_deps = []
 
             state.last_node = custom
             if isinstance(custom, GatherToLDS):
-                state.is_async = True
+                state.async_deps.append(node)
 
         if isinstance(custom, NestedRegionOp):
-            add_shared_memory_barriers(
-                trace, trace.get_subgraph(custom.subgraph_name), info
-            )
+            subgraph = trace.get_subgraph(custom.subgraph_name)
+            if not any(i.async_deps for i in info.values()):
+                add_shared_memory_barriers(trace, subgraph, info)
+                continue
+
+            first = get_first(subgraph.nodes)
+
+            # Convert dependencies to placeholders.
+            with subgraph.inserting_before(first):
```

Comment:
add tp util

Reply:
I've thought on it, but it too intertwined with the pass logic as we need to insert new ops and update the `SharedMemoryBarrierInfo` at the same time.